# STT 441: Probability, Section 002, Dr. Yuying Xie
## Homework 1 - 09/06/2024

This notebook contains the computational work completed by Lowell Monis toward Homework 1.

**Question 6.** There are 36 students in the classroom. Assuming each date of the year is equally likely to be the birthday of a student. Calculate the probability that at least two students have the same birthday, using simulation. 

Run your simulation 1,000,000 times and report how many times at least two students are having the same birthday.

***

### Solution

#### Setup

To tackle this question, one can utilize random simulations.

**Assumption:** Leap years do not exist in this experiment.

For this problem, I will be using Python. I will be dividing the code into three functions:

1. `generate_class()`: Returns a list representing a classroom with a random integer between 1 and 365 for each student, representing a day of the year. Accepts the number of students in the classroom.
2. `run_simulation()`: Runs a single simulation of the Birthday Problem. Accepts a list of the birthdays of the students. Uses the logic that if a list of unique birthdays is created from the input list, and if the length of the new list is less than the length of the original list, it indicates that at least two people shared the same birthday. Returns a Boolean value.
3. `run_experiment()`: Accepts the number of students in the classroom and the total number of simulations. Returns the probability by dividing total number of true results by the number of simulations conducted.

But before I proceed, I will import the `random` module to facilitate a fair experiment, where each day of the year is equally likely to be a person's birthday. I will also import the `numpy` module to generate the list of unique birthdays using `numpy.unique()`.

In [2]:
import random
import numpy as np

#### Function 1: `generate_class()`

In [16]:
def generate_class(n_students):
    class_birthdays = []                               # initializes empty list to store generated birthdays
    for i in range(n_students):                        # iterates through each student
        class_birthdays.append(random.randint(1, 365)) # appends a random birthday to the student
    return class_birthdays

#### Function 2: `run_simulation()`

In [15]:
def run_simulation(birthdays):
    unique_birthdays = np.unique(birthdays)            # creates a new list with the unique birthday from the original list
    if len(unique_birthdays) < len(birthdays):         # checks if the new list reduces in length to verify for shared birthdays and returns boolean
        return True
    else:
        return False

#### Function 3: `run_experiment()`

In [17]:
def run_experiment(n_simulations, n_students):
    count_true_scenarios = 0                           # initializing variable to store the number of true results
    for i in range(n_simulations):                     # iterating through each simulation
        if run_simulation(generate_class(n_students)): # generating a classroom and running the simulation on it
            count_true_scenarios += 1
    return count_true_scenarios/n_simulations          # returning probability

#### Calculating the probability

In [13]:
# This cell will take time to run.

number_of_students = 36
number_of_simulations = 1000000

probability = run_experiment(number_of_simulations, number_of_students)

print("The probability that at least two students share the same birthday in a classroom of", number_of_students, "in an experiment with", number_of_simulations, "is", probability)

The probability that at least two students share the same birthday in a classroom of 36 in an experiment with 1000000 is 0.832337


***

The probability is around 83% in each run of the experiment, thus imitating the results of the theoretical calculation. Hence, the simulations verify the result. (q.e.d.)

***